### DISCUSSION TITLE
Issue with the versioned_rows example from documentation

### DISCUSSION THREAD

**Discussion by davidlsq** [sqlalchemy/sqlalchemy#10995] [Usage Questions] [answered]
Hello,

I am using SQLAlchemy 2.0.16 and want to implement a versioned table as in https://docs.sqlalchemy.org/en/20/_modules/examples/versioned_rows/versioned_rows_w_versionid.html

But while testing the atomicity, I have a race condition :

```python
session = Session()
e = Example(id=2, data="A")
session.add(e)
session.commit()

session1 = Session()
e1 = session1.query(Example).filter_by(id=2).one()
e1.data = "B1"

session2 = Session()
e2 = session2.query(Example).filter_by(id=2).one()
e2.data = "B2"

session1.commit()
session2.commit()
```
```
sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: example.id, example.version_id
[SQL: INSERT INTO example (data, id, version_id, is_current_version) VALUES (?, ?, ?, ?)]
[parameters: ('B2', 2, 2, 1)]
```

I can fix it by replacing :
```python
self.version_id += 1
```
with :
```python
self.version_id = (
    select(func.max(self.__class__.version_id) + 1)
    .filter_by(id=old_id)
    .scalar_subquery()
)
```

It works but I have wrong partial updates :

```python
class Example(Versioned, Base):
    __tablename__ = "example"
    data1 = Column(String)
    data2 = Column(String)

Base.metadata.create_all(engine)

session = Session()
e = Example(id=2, data1="A", data2="B")
session.add(e)
session.commit()

session1 = Session()
e1 = session1.query(Example).filter_by(id=2).one()
e1.data1 = "A1"

session2 = Session()
e2 = session2.query(Example).filter_by(id=2).one()
e2.data2 = "B2"

session1.commit()
session2.commit()
```
```
id|version_id|is_current_version|data1|data2|
--+----------+------------------+-----+-----+
 2|         1|false             |A    |B    |
 2|         2|false             |A1   |B    |
 2|         3|true              |A    |B1   | <- instead of A1 B1
```

Is there a way to have a mixin with correct atomic versioning for the ORM ?

**Comment by zzzeek** [comment]
Hi -

unfortunately I dont have any particular insight in how to solve the problem in this case since if these were UPDATE statements, you would use the regular version_id approach and/or SELECT..FOR UPDATE to optimistically or pessimistically lock.  since these are INSERTs we lose those facilities.      I used this pattern in an application many years ago where we didnt have this kind of concurrency issue.

**Reply by davidlsq** [reply]
Yes, the original code works with a `FOR UPDATE` on postgres :
```python
with Session() as session:
    with session.begin():
        session.query(Example).filter_by(id=2, calc_is_current_version=True).with_for_update().first()
        e1 = session.query(Example).filter_by(id=2, calc_is_current_version=True).one()
        e1.data1 = "A1"
```

A bit frustrating to not be able to do with ORM what we can with a SQL query :
```sql
WITH new_version_id AS (
  INSERT INTO example (id, version_id, is_current_version, data1, data2)
  SELECT id, version_id + 1, true, 'A1', data2
  FROM example
  WHERE id = 2 AND is_current_version
  RETURNING version_id
)
UPDATE example
SET is_current_version = false
WHERE id = 2 AND version_id = (SELECT version_id FROM new_version_id) - 1
```
But if it keeps the code simple

Thanks

**Reply by zzzeek** [reply]
You can emit that SQL as written using a [bulk insert /update pattern](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#) to update your objects and you can also add RETURNING to get back new instances that match the rows just inserted.

you can always emit whatever SQL you want.  the ORM unit of work though only knows how to emit simple INSERT/UPDATE statements so you'd need to forego using flush, unless you wanted to do some elaborate SQL modification events which might not be very easy to do.   I would prevent the objects of this type from being flushed using the before_flush() event and you can potentially even emit the special CTE form you have above within before_flush().

**Reply by davidlsq** [reply]
> You can emit that SQL as written using a [bulk insert /update pattern](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#) to update your objects and you can also add RETURNING to get back new instances that match the rows just inserted.

I didn't manage to do it, I can't use the `RETURNING` as a subquery or am I missing something ?

The closer I get is :

```python
def update_data(session, id, data1, data2):
    new_version = session.scalar(
        insert(Example).values({
            "id": id,
            "version_id": select(Example.version_id + 1).filter_by(id=id, is_current_version=True).scalar_subquery(),
            "is_current_version": True,
            "data1": data1 if data1 else select(Example.data1).filter_by(id=id, is_current_version=True).scalar_subquery(),
            "data2": data2 if data2 else select(Example.data2).filter_by(id=id, is_current_version=True).scalar_subquery(),
        }).returning(Example.version_id)
    )
    session.execute(
        update(Example),
        [
            {
                "id": id,
                "version_id": new_version - 1,
                "is_current_version": False
            },
        ]
    )
```
But that gives the same race condition without a lock on the rows

I can avoid it by dropping the column `is_current_version` and use instead the calculated column `calc_is_current_version` :
```python
def update_data(session, id, data1, data2):
    session.execute(
        insert(Example).values({
            "id": id,
            "version_id": select(Example.version_id + 1).filter_by(id=id, calc_is_current_version=True),
            "data1": data1 if data1 else select(Example.data1).filter_by(id=id, calc_is_current_version=True),
            "data2": data2 if data2 else select(Example.data2).filter_by(id=id, calc_is_current_version=True),
        })
    )
```

> you can potentially even emit the special CTE form you have above within before_flush()

With my `insert` above, I need to programmatically know if `data1` and `data2` has changed. I have a working non-generic code for `Example` but not sure it's the best method :
```python
@event.listens_for(Session, "before_flush")
def before_flush(session, flush_context, instances):
    for instance in session.dirty:
        if not isinstance(instance, Example):
            continue
        ...
        data1_changed = inspect(instance).attrs.data1.history.has_changes()
        data2_changed = inspect(instance).attrs.data2.history.has_changes()
        session.execute(
            insert(Example).values({
                "id": id,
                "version_id": select(Example.version_id + 1).filter_by(id=id, calc_is_current_version=True),
                "data1": instance.data1 if data1_changed else select(Example.data1).filter_by(id=id, calc_is_current_version=True),
                "data2": instance.data2 if data2_changed else select(Example.data2).filter_by(id=id, calc_is_current_version=True),
            })
        )
```
I can probably make it generic with some more inspection to get the columns programmatically

**Reply by zzzeek** [reply, marked as answer]
I dont yet understand how this query prevents the condition you show originally which is this:

```
id|version_id|is_current_version|data1|data2|
--+----------+------------------+-----+-----+
 2|         1|false             |A    |B    |
 2|         2|false             |A1   |B    |
 2|         3|true              |A    |B1   | <- instead of A1 B1
```

e1 and e2 both see the row with A and B and will still both emit their INSERT statements independently, the latest version still not having A1 B1.     if the UPDATE is being used here as a "lock", you can still do a SELECT..FOR UPDATE on the row, it will be locked, updates will not be allowed by other transactions, the regular recipe will still work then, no ?    this is the usual way to handle this problem.

re: RETURNING, I meant you would have the RETURNING on the outer UPDATE where it would return columns from the CTE.

**Reply by davidlsq** [reply]
Ah my bad I was wrong on the race condition, from my test in my first message I thought that it was safe without a lock and only :
```python
self.version_id = (
    select(func.max(self.__class__.version_id) + 1)
    .filter_by(id=old_id)
    .scalar_subquery()
)
```
But it's actually not

I understand now that I have to lock to make it work

Thanks a lot for your help 🙏
